### PR TITLE
#10 add fixtures

### LINKED ISSUE #10: Add fixtures to pytest_saas
Prerequisites 
* [x] Create a plugin, see https://github.com/exasol/pytest-plugins/issues/9

### Support providing a short name for the current project
Candidates for projects: 
* saas-api-python 
* bucketfd-python

Add support reading the project short tag from different sources with precedence as ordered
* [x] Support CLI option `--project-short-tag`
* [x] Support Environment variable `PROJECT_SHORT_TAG`
* [x] Read project short tag from yaml file `error_code_config.yml`

### Content of file `test/test/conftest.py`

```python
pytest_plugins = [  "fixtures.db_name", ] 
```

### Content of file `test/fixtures/db_name.py`

See
* [pytest_configure](https://docs.pytest.org/en/8.2.x/how-to/writing_plugins.html#registering-custom-markers)
* [rootpath](https://docs.pytest.org/en/6.2.x/reference.html?highlight=config#pytest.config.Config.rootpath)

```python
def pytest_add_options()
   # support cli options to specify project name 
   ....
   
def pytest_configure(config)
   if error_code_config exists
        # maybe config.rootpath is helpful?
        config.stash["db_name"] = read_short_tag(os.cwd)
   elif cli option set:
        config.stash["db_name"] = read_cli_option()
   elif environment variable set:
        config.stash["db_name"] = read_environment_variable()
   elif ini value set:
        config.stash["db_name"] = read_ini_value()     
   else:
        default = "none"
    
@fixture
def db_name(pytestconfig)
    return pytestconfig.stash["db_name"]
```

### PR DESCRIPTION
**Pull request by ckunki**
closes #10

### CODE REVIEW

File: pytest-saas/README.md [hunk 1]
```diff
@@ -15,3 +15,43 @@ To install the pytest-exasol-saas plugin, you can use pip:
...
+pytest-exasol-saas Plugin will name the database instances using 3 components
+
+* **Project Short Tag**: Abbreviation of the current project, see below for different [options for providing the project short tag](#options-for-providing-the-project-short-tag).
+* **Timestamp**: Number of seconds since epoc, see [Unix Time](https://en.wikipedia.org/wiki/Unix_time).
+* A dash character `-`
+* **User Name**: Login name of the current user.
+
+A database instances might for example have the name `1715155224SAPIPY-run` indicating it was
+* created on Wednesday, May 8, 2024
+* in the context of a project with short tag `SAPYPI`
```
Comment (tkilias) [inline pytest-saas/README.md:46]:
```suggestion
* in the context of a project with short tag `SAPIPY`
```

File: pytest-saas/exasol/pytest_saas/project_short_tag.py [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+    """
+    Read project-short-tag from yaml file ``FILE`` in the specified
+    directory ``dir``.
+    """
+    config_file = dir / FILE
+    if not config_file.exists():
+        return None
+    content = config_file.read_text()
+    header = False
+    for line in content.splitlines():
```
Comment (tkilias) [inline pytest-saas/exasol/pytest_saas/project_short_tag.py:24]: I had guessed the error-code file is json or yaml, in that case would have parsed it that way. The would be much more reliable.

Comment (tkilias) [reply pytest-saas/exasol/pytest_saas/project_short_tag.py:24]: you can use pyyaml to read yaml files

File: pytest-saas/test/integration/pytest_saas_test.py [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+              assert request.config.getoption("--keep-saas-database")
+              assert "123" == request.config.getoption("--saas-database-id")
+              assert "PST" == request.config.getoption("--project-short-tag")
+          """),
+          _cli_args(
+              "--keep-saas-database",
+              "--project-short-tag", "PST",
+              "--saas-database-id", "123",
+          ),
+         ),
```
Comment (tkilias) [inline pytest-saas/test/integration/pytest_saas_test.py:54]: What is the intention of these tests. Do want to test, if the CLI options are defined?

File: pytest-saas/test/integration/pytest_saas_test.py
Comment (tkilias) [inline pytest-saas/test/integration/pytest_saas_test.py:1]: I think, we are missing a test that actual tests if our fixtures work. For example, one which creates an actual SaaS DB and uses it somehow in the test.

**Review by tkilias — CHANGES_REQUESTED** [review]

File: pytest-saas/test/integration/pytest_saas_test.py (hunk 3 above)
Comment (ckunki) [reply pytest-saas/test/integration/pytest_saas_test.py:54]: Yes. I took a look at the pytest-exasol-itde plugin and created similar test cases.
This is also based on a discussion with @Nicoretti.

File: pytest-saas/exasol/pytest_saas/project_short_tag.py (hunk 2 above)
Comment (ckunki) [reply pytest-saas/exasol/pytest_saas/project_short_tag.py]: see next push

File: pytest-saas/test/integration/pytest_saas_test.py (hunk 3 above)
Comment (ckunki) [reply pytest-saas/test/integration/pytest_saas_test.py:54]: see next push

File: justfile [hunk 4]
```diff
@@ -6,11 +6,18 @@ default:
 
 # Run tests for one or multiple projects within this respository
 test +projects=PROJECTS:
-    #!/usr/bin/env bash
-    for p in {{projects}}; do
-        poetry -C ${p}/ install
-        poetry -C ${p}/ run nox -f ${p}/noxfile.py -s coverage
-    done
+    #!/usr/bin/env python
```
Comment (tkilias) [inline justfile:9]:
```suggestion
    #!/usr/bin/env python3
```

File: justfile [hunk 5]
```diff
@@ -6,11 +6,18 @@ default:
...
+    import subprocess, sys
+    rc = 0
+    def run(command):
+        global rc
```
Comment (tkilias) [inline justfile:13]: A bit hacky, but efficient

File: pytest-saas/README.md [hunk 6]
```diff
@@ -15,3 +15,43 @@ To install the pytest-exasol-saas plugin, you can use pip:
...
 pip install pytest-exasol-saas
 ```
+
+## Database Instances
+
+### Using an Existing Database Instance
+
+By default the fixtures in pytest-exasol-saas Plugin will create instances of Exasol SaaS database with scope `session`.
+
+If you want to use an existing instance instead, then you can provide the instance's ID with command line option `--saas-database-id <ID>` to pytest.
```
Comment (tkilias) [inline pytest-saas/README.md:25]:
```suggestion
If you want to use an existing instance instead, then you can provide the instance's ID with the command line option `--saas-database-id <ID>` to pytest.
```

File: pytest-saas/README.md [hunk 7]
```diff
@@ -15,3 +15,43 @@ To install the pytest-exasol-saas plugin, you can use pip:
...
+
+### Keeping Database Instances After the Test Session
+
+By default the fixtures in pytest-exasol-saas Plugin will remove the created database instances after the session or in case of errors.
+
+However, if you provide command line option `--keep-saas-database` then pytest-exasol-saas Plugin will _keep_ these instances for subsequent inspection or reuse.
```
Comment (tkilias) [inline pytest-saas/README.md:31]:
```suggestion
However, if you provide the command line option `--keep-saas-database` then pytest-exasol-saas Plugin will _keep_ these instances for subsequent inspection or reuse.
```

File: README.md [hunk 8]
```diff
@@ -0,0 +1,62 @@
...
+Replace x.y.z with the desired version number.
+
+## Development
+
+Before you can start developing in this workspace, please ensure you have the following tools installed either globally or at a workspace level.
+
+* [Python](https://www.python.org)
+* [Just](https://github.com/casey/just)
+
+## Run Tests
```
Comment (tkilias) [inline README.md:49]: This needs to go to the Developer Guide

File: README.md [hunk 9]
```diff
@@ -0,0 +1,62 @@
...
+```shell
+pip install pytest-itde
+```
+
+To install a specific version of a plugin, simply specify the version number:
+
+```shell
+pip install "pytest-itde==x.y.z"
+```
+
```
Comment (tkilias) [inline README.md:39]: Installation of pytest-exasol-saas is missing

File: README.md [hunk 10]
```diff
@@ -0,0 +1,62 @@
...
+## Installation
+
+To ensure you're using the latest features and bug fixes, we recommend installing the plugins directly from PyPI using your preferred package manager. This approach simplifies the process of keeping your testing environment up-to-date.
+
+For example, to install the `pytest-itde` plugin, you could use the following command:
+
+
+```shell
+pip install pytest-itde
```
Comment (tkilias) [inline README.md:31]: Inconsistent with the Plugins section

Comment (ckunki) [reply README.md:31]: Fixed in next push

File: README.md (hunk 9 above)
Comment (ckunki) [reply README.md:39]: This is the overall README of the repository pytest-plugins. 
The README says 
> For example, to install the `pytest-exasol-itde` plugin, you could use the following command:

I propose to leave the example as it and to not duplicate installations of all potential different plugins in their own resp. README files.

File: README.md (hunk 8 above)
Comment (ckunki) [reply README.md:49]: See next push

File: README.md [hunk 11]
```diff
@@ -0,0 +1,62 @@
...
+
+To install a specific version of a plugin, simply specify the version number:
+
+```shell
+pip install "pytest-itde==x.y.z"
+```
+
+Replace x.y.z with the desired version number.
+
+## Development
```
[the remark this answers is not in the document]
Comment (ckunki) [reply README.md:42]: See next push

**Review by tkilias — APPROVED** [review]

**Pull request merged**